README.md
[![pub package](https://img.shields.io/pub/v/smoothie.svg)](https://pub.dartlang.org/packages/smoothie)

# smoothie

Create and sample a smooth Bézier curve going through a given set of points

## Usage

Let's say you have some arbitrary data (a list of points):

```dart
var _originalDataSeries = <Point>[
    Point(0, 5),
    Point(3, 15),
    Point(5, 10),
    Point(6, 6),
    Point(9, 13),
  ];
```

Define a desired number of samples in the resulting curve:

```dart
int _sampleCount = 100;
```
Import `smoothie`:

```
import 'package:smoothie/smoothie.dart';
```

Create a smooth curve using the `getSampledCurveFromPoints` extension function on your original data:

```dart
List<Point<num>> _sampledCurve = _originalDataSeries.smooth(_sampleCount);
```

See oversampling in action:

<img src="https://raw.githubusercontent.com/alekskuzmin/smoothie/master/example/example.gif" width="270" height="480">

Full example using the `charts_flutter` package:

```dart
import 'dart:math';

import 'package:charts_flutter/flutter.dart' as charts;
import 'package:flutter/material.dart';
import 'package:smoothie/smoothie.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Smoothie Demo',
      theme: ThemeData(
        primarySwatch: Colors.teal,
      ),
      home: SmoothieHomePage(),
    );
  }
}

class SmoothieHomePage extends StatefulWidget {
  SmoothieHomePage({Key? key}) : super(key: key);

  @override
  _SmoothieHomePageState createState() => _SmoothieHomePageState();
}

class _SmoothieHomePageState extends State<SmoothieHomePage> {
  int _samplingPointCount = 0;

  var _originalDataSeries = <Point>[
    Point(0, 5),
    Point(2, 15),
    Point(3, 10),
    Point(8, 6),
    Point(9, 13),
  ];

  @override
  void initState() {
    super.initState();
    _samplingPointCount = _originalDataSeries.length;
  }

  void _incrementCounter() {
    setState(() {
      _samplingPointCount++;
    });
  }

  void _decrementCounter() {
    setState(() {
      if (_samplingPointCount > _originalDataSeries.length)
        _samplingPointCount--;
    });
  }

  @override
  Widget build(BuildContext context) {
    var series = [
      new charts.Series(
        domainFn: (Point chartData, _) => chartData.x,
        measureFn: (Point chartData, _) => chartData.y,
        colorFn: (Point point, _) => charts.MaterialPalette.teal.shadeDefault,
        id: 'Example Series',
        data: _originalDataSeries.smooth(
          _samplingPointCount,
        ),
      ),
    ];

    var chart = new charts.LineChart(
      series,
      animate: false,
      defaultRenderer: new charts.LineRendererConfig(
        includeArea: true,
      ),
    );

    return Scaffold(
        appBar: AppBar(
          title: Text("Smoothie Demo"),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Text(
                'The smooth curve now has $_samplingPointCount points.',
              ),
              Container(
                height: 200,
                child: chart,
              ),
            ],
          ),
        ),
        floatingActionButton: Row(
          mainAxisAlignment: MainAxisAlignment.end,
          children: <Widget>[
            FloatingActionButton(
              onPressed: _decrementCounter,
              tooltip: 'Decrement',
              child: Icon(Icons.remove),
            ),
            FloatingActionButton(
              onPressed: _incrementCounter,
              tooltip: 'Increment',
              child: Icon(Icons.add),
            ),
          ],
        ));
  }
}
```

If you want to run the example as is and you are using null safety, you have to pass `--no-sound-null-safety` to the build command as the current version of `charts_flutter` does not yet support null safety:
```console
cd example
flutter run --no-sound-null-safety
```



CHANGELOG.md
## [0.3.0] - 2021-04-09

### Null safety
The package is now null-safe. Added a static function in a Smoothie class for convenient import.

## [0.2.0+1] - 2020-05-12

### Breaking changes
Re-worked the API for smoothing function. The new function is designed to handle input data with points arbitrarily (not evenly) spread across X axis.

## [0.1.4] - 2020-05-12

### Updated example

## [0.1.3] - 2020-05-12

### Improved documentation

## [0.1.2] - 2020-05-04

### Updated the example

## [0.1.1] - 2020-05-04

### Separate extension for list of custom data points

## [0.1.0] - 2020-05-04

### Breaking change - using extension methods

## [0.0.4] - 2020-05-02

### Improved naming and documentation

## [0.0.3] - 2020-04-23

### Improved Readme example

## [0.0.2] - 2020-04-01

### Added the example app

## [0.0.1] - 2020-03-31

### Initial release

example/smoothie_example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:smoothie_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/smoothie.dart
///Create and sample a smooth Bézier curve going through a given set of points
library smoothie;

import 'dart:math';

import 'package:bezier/bezier.dart';
import 'package:vector_math/vector_math.dart';

class Smoothie {
  ///Creates a sampled curve going through list of points.
  ///
  ///[sampleCount] is a number of points of the resulting smoothed curve
  ///
  ///Returns a list of length [sampleCount] points that lay on a curve going through the given points.
  static List<Point<num>> smooth(
    List<Point> points,
    num sampleCount, {
    bool isSorted = true,
  }) {
    return points.smooth(sampleCount, isSorted: isSorted);
  }
}

extension SmoothPoints on List<Point> {
  ///Creates a sampled curve going through list of points.
  ///
  ///[sampleCount] is a number of points of the resulting smoothed curve
  ///
  ///Returns a list of length [sampleCount] points that lay on a curve going through the given points.
  List<Point<num>> smooth(
    num sampleCount, {
    bool isSorted = true,
  }) {
    if (!isSorted) {
      this.sort();
    }
    var sampleWidth = (this[this.length - 1].x - this[0].x) / sampleCount;
    List<Point> sampledCurve = [];
    for (int originalPointIndex = 0;
        originalPointIndex < this.length - 1;
        originalPointIndex++) {
      var currentIntervalWidth =
          (this[originalPointIndex + 1].x - this[originalPointIndex].x);
      int intervalSampleCount =
          max(1, (currentIntervalWidth / sampleWidth).round());
      var curve = new CubicBezier(
        [
          Vector2(0, this[originalPointIndex].y.toDouble()),
          Vector2(intervalSampleCount.toDouble() / 2,
              this[originalPointIndex].y.toDouble()),
          Vector2(intervalSampleCount.toDouble() / 2,
              this[originalPointIndex + 1].y.toDouble()),
          Vector2(intervalSampleCount.toDouble(),
              this[originalPointIndex + 1].y.toDouble()),
        ],
      );
      for (int oversamplingIndex = 0;
          oversamplingIndex <= intervalSampleCount;
          oversamplingIndex++) {
        sampledCurve.add(Point(
            (this[originalPointIndex].x +
                    currentIntervalWidth *
                        oversamplingIndex /
                        intervalSampleCount)
                .toDouble(),
            curve.pointAt(
                (oversamplingIndex / intervalSampleCount).toDouble())[1]));
      }
    }
    return sampledCurve;
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:charts_flutter/flutter.dart' as charts;
import 'package:flutter/material.dart';
import 'package:smoothie/smoothie.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Smoothie Demo',
      theme: ThemeData(
        primarySwatch: Colors.teal,
      ),
      home: SmoothieHomePage(),
    );
  }
}

class SmoothieHomePage extends StatefulWidget {
  SmoothieHomePage({Key? key}) : super(key: key);

  @override
  _SmoothieHomePageState createState() => _SmoothieHomePageState();
}

class _SmoothieHomePageState extends State<SmoothieHomePage> {
  int _samplingPointCount = 0;

  var _originalDataSeries = <Point>[
    Point(0, 5),
    Point(2, 15),
    Point(3, 10),
    Point(8, 6),
    Point(9, 13),
  ];

  @override
  void initState() {
    super.initState();
    _samplingPointCount = _originalDataSeries.length;
  }

  void _incrementCounter() {
    setState(() {
      _samplingPointCount++;
    });
  }

  void _decrementCounter() {
    setState(() {
      if (_samplingPointCount > _originalDataSeries.length)
        _samplingPointCount--;
    });
  }

  @override
  Widget build(BuildContext context) {
    var series = [
      new charts.Series(
        domainFn: (Point chartData, _) => chartData.x,
        measureFn: (Point chartData, _) => chartData.y,
        colorFn: (Point point, _) => charts.MaterialPalette.teal.shadeDefault,
        id: 'Example Series',
        data: _originalDataSeries.smooth(
          _samplingPointCount,
        ),
      ),
    ];

    var chart = new charts.LineChart(
      series,
      animate: false,
      defaultRenderer: new charts.LineRendererConfig(
        includeArea: true,
      ),
    );

    return Scaffold(
        appBar: AppBar(
          title: Text("Smoothie Demo"),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Text(
                'The smooth curve now has $_samplingPointCount points.',
              ),
              Container(
                height: 200,
                child: chart,
              ),
            ],
          ),
        ),
        floatingActionButton: Row(
          mainAxisAlignment: MainAxisAlignment.end,
          children: <Widget>[
            FloatingActionButton(
              onPressed: _decrementCounter,
              tooltip: 'Decrement',
              child: Icon(Icons.remove),
            ),
            FloatingActionButton(
              onPressed: _incrementCounter,
              tooltip: 'Increment',
              child: Icon(Icons.add),
            ),
          ],
        ));
  }
}
